api/data_explorer/dataset_faceted_search.py

```diff
@@ -1,40 +1,78 @@
+from elasticsearch_dsl import Mapping
 from elasticsearch_dsl import TermsFacet
 
 
-def _get_index_name():
-    """Gets index name from /app/dataset.json."""
-    with open('/app/dataset.json') as f:
+def get_index_name():
+    """Gets index name from dataset.json."""
+    with open('/app/config/dataset.json') as f:
         # Remove comments using jsmin, as recommended by JSON creator
         # (https://plus.google.com/+DouglasCrockfordEsq/posts/RK8qyGVaGSr).
         dataset = json.loads(jsmin.jsmin(f.read()))
         return dataset['name']
 
 
+def get_facets():
+    """Gets facets from facet_fields.csv.
+
+    app.config['ELASTICSEARCH_URL'], app.config['INDEX_NAME'] must be set before
+    this is called.
+    """
+    f = open('/app/config/facet_fields.csv')
+    # Remove comments using jsmin.
+    csv_str = jsmin.jsmin(f.read())
+    facet_rows = csv.DictReader(iter(csv_str.splitlines()), skipinitialspace=True)
+
+    using = Elasticsearch(current_app.config['ELASTICSEARCH_URL'])
+    mapping = Mapping.from_es(current_app.config['INDEX_NAME'], 'type',
+            using=using).to_dict()
+    facets = {}
+    for facet_row in facet_rows:
+        field_name = facet_row['readable_field_name']
+        field_type = mapping['type']['properties'][field_name]['type']
+        if field_type == 'text':
+            # Use ".keyword" because we want aggregation on keyword field, not
+            # term field. See
+            # https://www.elastic.co/guide/en/elasticsearch/reference/6.2/fielddata.html#before-enabling-fielddata
+            facets[field_name] = TermsFacet(field=field_name + '.keyword')
+        else:
+            # Assume numeric type.
+            # TODO: Handle other types.
+            facets[field_name] = HistogramFacet(field=field_name, interval=7)
+    current_app.logger.info('dataset_faceted_search facets: %s' % facets)
+    return facets
+
+
 class DatasetFacetedSearch(FacetedSearch):
     """Subclass of FacetedSearch for Datasets.
 
-    This class body is run when module is loaded; there is no app context yet.
-    App context contains configuration such as Elasticsearch URL. To use this
-    class, specify Elasticsearch URL like so:
-
-        search = DatasetFacetedSearch()
-        search.using = Elasticsearch(current_app.config['ELASTICSEARCH_URL'])
-        # Need to rebuild search._s with new s.using.
-        search.__init__()
-        response = search.execute()
+    app.config['ELASTICSEARCH_URL'], app.config['INDEX_NAME'], and
+    app.config['FACETS'] must be set before creating a DatasetFacetedSearch
+    object.
     """
-    index = _get_index_name()
-    # TODO(melissachang): Get facets from facet_fields.csv.
-    facets = {
-        'Gender': TermsFacet(field='Gender.keyword'),
-        'Region': TermsFacet(field='Region.keyword'),
-    }
+
+    def __init__(self):
+        """Initializes DatasetFacetedSearch object.
+
+        Contains initialization that should be done once over the lifetime of
```

Comment:
Same as my comment above, people reading the Python here are probably familiar with `__init__(self)`.

Reply:
Actually this comment is wrong, I moved the once-per-server initialization to to __main__ init(). Deleted comment.